File: app/src/main/java/com/gmail/mostafa/ma/saleh/yawa/retrofit/OnFinishedListener.java
package com.gmail.mostafa.ma.saleh.yawa.retrofit;

import android.support.annotation.CallSuper;
import android.support.annotation.Nullable;

/**
 * Created by Mostafa on 10/23/2017.
 */

public abstract class OnFinishedListener<T> {

    private boolean cancelled;

    public void onSuccess(T arg) {
    }

    public void onFailure(String message) {
    }

    @CallSuper
    public void onComplete(@Nullable T args, @Nullable String message) {
        if (!cancelled) {
            if (args != null) {
                onSuccess(args);
            } else {
                onFailure(message);
            }
        }
    }

    public void cancel() {
        cancelled = true;
    }

    public boolean isCancelled() {
        return cancelled;
    }
}

File: app/src/main/java/com/gmail/mostafa/ma/saleh/yawa/retrofit/OnFinishedListener.kt
package com.gmail.mostafa.ma.saleh.yawa.retrofit

import android.support.annotation.CallSuper

/**
 * Created by Mostafa on 10/23/2017.
 */

abstract class OnFinishedListener<T> {

    var isCancelled: Boolean = false
        private set

    open fun onSuccess(arg: T) {}

    open fun onFailure(message: String?) {}

    @CallSuper
    open fun onComplete(args: T?, message: String?) {
        if (!isCancelled) {
            if (args != null) {
                onSuccess(args)
            } else {
                onFailure(message)
            }
        }
    }

    fun cancel() {
        isCancelled = true
    }
}
